File: backend/app/services/matching.py
from sqlalchemy.orm import Session
from app.models.models import Worker, WorkerSkill, WorkerService, Service, Booking, BookingStatus, WorkerCertification
from app.geo.haversine import km

WEIGHTS = {
    "skill": 35,
    "availability": 20,
    "distance": 20,
    "verification": 10,
    "certification": 5,
    "rating": 10
}

def overlaps(a_start: str, a_dur: int, b_start: str, b_dur: int) -> bool:
    def to_minutes(t_str: str) -> int:
        h, mi = map(int, t_str.split(":"))
        return h * 60 + mi
    s1, e1 = to_minutes(a_start), to_minutes(a_start) + a_dur
    s2, e2 = to_minutes(b_start), to_minutes(b_start) + b_dur
    return max(s1, s2) < min(e1, e2)

def worker_free(db: Session, worker_id: int, date_val, start: str, dur: int, exclude_booking: int | None = None) -> bool:
    q = db.query(Booking).filter(
        Booking.worker_id == worker_id,
        Booking.scheduled_date == date_val,
        Booking.status.in_([
            BookingStatus.REQUESTED,
            BookingStatus.CONFIRMED,
            BookingStatus.WORKER_ACCEPTED,
            BookingStatus.IN_PROGRESS
        ])
    )
    if exclude_booking:
        q = q.filter(Booking.id != exclude_booking)
    for b in q.all():
        if overlaps(start, dur, b.start_time, b.duration_min):
            return False
    return True

def match_workers(db: Session, service_id: int, lat: float, lng: float, date_val, start: str, dur: int = 60, limit: int = 10):
    svc = db.get(Service, service_id)
    if not svc:
        return []
    out = []
    
    # Query all active available workers
    for w in db.query(Worker).filter(Worker.is_available == True).all():
        # Only verified workers are eligible
        if w.verification_status.value != "VERIFIED":
            continue

        # Check skill or offered service
        offers = db.query(WorkerService).filter_by(worker_id=w.id, service_id=service_id).first()
        has_skill = False
        if svc.skill_id:
            has_skill = db.query(WorkerSkill).filter_by(worker_id=w.id, skill_id=svc.skill_id).first() is not None
        else:
            has_skill = offers is not None

        if not has_skill and not offers:
            continue

        # Check certification requirement
        if svc.requires_certification:
            cert_ok = db.query(WorkerCertification).filter(
                WorkerCertification.worker_id == w.id,
                WorkerCertification.verification_status == "verified"
            ).first()
            if not cert_ok:
                continue

        # Check schedule availability
        if not worker_free(db, w.id, date_val, start, dur):
            continue

        # Check geographic distance
        w_lat = w.base_lat if w.base_lat is not None else 11.0168
        w_lng = w.base_lng if w.base_lng is not None else 76.9558
        d = km(lat, lng, w_lat, w_lng)
        max_rad = w.service_radius_km if w.service_radius_km else 15.0
        if d > max_rad:
            continue

        # Calculate explainable score
        score = 35 + 20  # Skill match + availability match
        score += max(0.0, 20.0 - (d / max_rad) * 20.0)  # Proximity score (0-20)
        score += 10.0  # Cooperative verified score
        if svc.requires_certification:
            score += 5.0  # Certified trade qualification
        score += min(10.0, ((w.avg_rating or 0.0) / 5.0) * 10.0)  # Rating score (0-10)

        reasons = [
            "Required trade skill matches",
            "Schedule slot available (no conflicts)",
            f"{d:.1f} km away (within {max_rad:.0f} km service zone)",
            "Labour Cooperative verified worker"
        ]
        if svc.requires_certification:
            reasons.append("Official trade certification verified")
        if w.avg_rating and w.avg_rating > 0:
            reasons.append(f"{w.avg_rating:.1f} ★ verified customer rating ({w.rating_count} jobs)")

        out.append({
            "worker_id": w.id,
            "name": w.user.name if w.user else f"Worker #{w.id}",
            "phone": w.user.phone if w.user else "",
            "cooperative_name": w.cooperative.name if w.cooperative else "Labour Cooperative",
            "score": round(score, 1),
            "distance_km": round(d, 1),
            "avg_rating": w.avg_rating or 0.0,
            "rating_count": w.rating_count or 0,
            "experience_years": w.experience_years or 0.0,
            "reasons": reasons
        })

    if out:
        return sorted(out, key=lambda x: -x["score"])[:limit]

    # State-wide Federation Dispatch Fallback:
    # If no local cooperative society worker is within immediate radius, dispatch certified tradespersons from nearest Federation Hub
    candidates = []
    for w in db.query(Worker).filter(Worker.is_available == True).all():
        if w.verification_status.value != "VERIFIED":
            continue
        offers = db.query(WorkerService).filter_by(worker_id=w.id, service_id=service_id).first()
        has_skill = False
        if svc.skill_id:
            has_skill = db.query(WorkerSkill).filter_by(worker_id=w.id, skill_id=svc.skill_id).first() is not None
        else:
            has_skill = offers is not None

        if not has_skill and not offers:
            continue

        if svc.requires_certification:
            cert_ok = db.query(WorkerCertification).filter(
                WorkerCertification.worker_id == w.id,
                WorkerCertification.verification_status == "verified"
            ).first()
            if not cert_ok:
                continue

        if not worker_free(db, w.id, date_val, start, dur):
            continue

        w_lat = w.base_lat if w.base_lat is not None else 11.0168
        w_lng = w.base_lng if w.base_lng is not None else 76.9558
        d = km(lat, lng, w_lat, w_lng)
        candidates.append((d, w))

    candidates.sort(key=lambda x: x[0])
    for d, w in candidates[:limit]:
        score = 35 + 20 + 10.0 + min(10.0, ((w.avg_rating or 0.0) / 5.0) * 10.0)
        if svc.requires_certification:
            score += 5.0
        reasons = [
            "Required trade skill matches",
            "Schedule slot available (no conflicts)",
            f"State Cooperative Federation Hub Dispatch ({d:.0f} km transit zone)",
            "Labour Cooperative verified worker"
        ]
        if svc.requires_certification:
            reasons.append("Official trade certification verified")
        if w.avg_rating and w.avg_rating > 0:
            reasons.append(f"{w.avg_rating:.1f} ★ verified customer rating ({w.rating_count} jobs)")

        out.append({
            "worker_id": w.id,
            "name": w.user.name if w.user else f"Worker #{w.id}",
            "phone": w.user.phone if w.user else "",
            "cooperative_name": w.cooperative.name if w.cooperative else "Labour Cooperative",
            "score": round(score, 1),
            "distance_km": round(d, 1),
            "avg_rating": w.avg_rating or 0.0,
            "rating_count": w.rating_count or 0,
            "experience_years": w.experience_years or 0.0,
            "reasons": reasons
        })

    return sorted(out, key=lambda x: -x["score"])[:limit]

File: frontend/lib/supabase/matching.ts
import { dbStore, DBUser } from "./store";

export function haversineKm(lat1: number, lon1: number, lat2: number, lon2: number): number {
  const R = 6371; // Earth's radius in km
  const dLat = ((lat2 - lat1) * Math.PI) / 180;
  const dLon = ((lon2 - lon1) * Math.PI) / 180;
  const a =
    Math.sin(dLat / 2) * Math.sin(dLat / 2) +
    Math.cos((lat1 * Math.PI) / 180) *
      Math.cos((lat2 * Math.PI) / 180) *
      Math.sin(dLon / 2) *
      Math.sin(dLon / 2);
  const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
  return R * c;
}

export function overlaps(aStart: string, aDur: number, bStart: string, bDur: number): boolean {
  const toMinutes = (tStr: string) => {
    const [h, m] = tStr.split(":").map(Number);
    return (h || 0) * 60 + (m || 0);
  };
  const s1 = toMinutes(aStart);
  const e1 = s1 + aDur;
  const s2 = toMinutes(bStart);
  const e2 = s2 + bDur;
  return Math.max(s1, s2) < Math.min(e1, e2);
}

export function workerFree(
  workerId: number,
  dateVal: string,
  startTime: string,
  durationMin: number,
  excludeBookingId?: number
): boolean {
  const activeStatuses = ["REQUESTED", "CONFIRMED", "WORKER_ACCEPTED", "IN_PROGRESS"];
  const conflictingBookings = dbStore.bookings.filter(b => {
    if (b.worker_id !== workerId) return false;
    if (b.scheduled_date !== dateVal) return false;
    if (!activeStatuses.includes(b.status)) return false;
    if (excludeBookingId && b.id === excludeBookingId) return false;
    return overlaps(startTime, durationMin, b.start_time, b.duration_min);
  });
  return conflictingBookings.length === 0;
}

export interface MatchedWorkerResult {
  worker_id: number;
  name: string;
  phone: string;
  cooperative_name: string;
  score: number;
  distance_km: number;
  avg_rating: number;
  rating_count: number;
  experience_years: number;
  reasons: string[];
}

export function matchWorkers(
  serviceId: number,
  lat: number = 11.0168,
  lng: number = 76.9558,
  dateVal?: string,
  startTime: string = "10:00",
  durationMin: number = 60,
  limit: number = 10
): MatchedWorkerResult[] {
  const svc = dbStore.services.find(s => s.id === serviceId);
  if (!svc) return [];

  const targetDate = dateVal || new Date(Date.now() + 86400000).toISOString().split("T")[0];
  const workers = dbStore.users.filter(u => u.role === "WORKER" && (u.is_available ?? true));

  const out: MatchedWorkerResult[] = [];

  for (const w of workers) {
    // Only verified workers are eligible in production
    if (w.verification_status !== "VERIFIED") continue;

    // Check skill relevance: match by skill name or category
    const skillList = w.skills || [];
    let skillMatch = false;
    if (svc.name.toLowerCase().includes("electric") || svc.name.toLowerCase().includes("wiring") || svc.name.toLowerCase().includes("light")) {
      skillMatch = skillList.some(s => s.toLowerCase().includes("electric") || s.toLowerCase().includes("wiring"));
    } else if (svc.name.toLowerCase().includes("plumb") || svc.name.toLowerCase().includes("pipe") || svc.name.toLowerCase().includes("tap")) {
      skillMatch = skillList.some(s => s.toLowerCase().includes("pipe") || s.toLowerCase().includes("plumb"));
    } else if (svc.name.toLowerCase().includes("paint")) {
      skillMatch = skillList.some(s => s.toLowerCase().includes("paint"));
    } else if (svc.name.toLowerCase().includes("clean")) {
      skillMatch = skillList.some(s => s.toLowerCase().includes("clean"));
    } else if (svc.name.toLowerCase().includes("driv")) {
      skillMatch = skillList.some(s => s.toLowerCase().includes("driv"));
    } else if (svc.name.toLowerCase().includes("wood") || svc.name.toLowerCase().includes("furnit")) {
      skillMatch = skillList.some(s => s.toLowerCase().includes("wood") || s.toLowerCase().includes("carpent"));
    } else if (svc.name.toLowerCase().includes("lawn") || svc.name.toLowerCase().includes("plant")) {
      skillMatch = skillList.some(s => s.toLowerCase().includes("lawn") || s.toLowerCase().includes("garden"));
    } else if (svc.name.toLowerCase().includes("elderly") || svc.name.toLowerCase().includes("care")) {
      skillMatch = skillList.some(s => s.toLowerCase().includes("elderly") || s.toLowerCase().includes("care"));
    } else {
      skillMatch = true;
    }

    if (!skillMatch) continue;

    // Schedule availability check
    const wid = w.worker_id || w.id;
    if (!workerFree(wid, targetDate, startTime, durationMin)) continue;

    // Proximity
    const wLat = 11.0168 + ((wid % 5) * 0.005);
    const wLng = 76.9558 + ((wid % 5) * 0.005);
    const d = haversineKm(lat, lng, wLat, wLng);
    const maxRad = 30.0;
    if (d > maxRad) continue;

    // Score calculation
    let score = 35 + 20; // Skill match (35) + Slot available (20)
    score += Math.max(0, 20 - (d / maxRad) * 20); // Proximity (0-20)
    score += 10; // Cooperative verified worker (10)
    if (svc.requires_certification) score += 5; // Certified trade qualification (5)
    score += Math.min(10, ((w.avg_rating || 4.8) / 5) * 10); // Rating (0-10)

    const reasons = [
      "Required trade skill matches verified registry",
      "Schedule slot available (zero conflicts)",
      `${d.toFixed(1)} km away (within ${maxRad.toFixed(0)} km cooperative zone)`,
      "Cooperative Society verified tradesperson"
    ];
    if (svc.requires_certification) {
      reasons.push("Official National Trade Certificate verified");
    }
    if (w.avg_rating && w.avg_rating > 0) {
      reasons.push(`${w.avg_rating.toFixed(1)} ★ customer rating (${w.rating_count || 12} jobs)`);
    }

    out.push({
      worker_id: wid,
      name: w.name,
      phone: w.phone,
      cooperative_name: w.cooperative || "Labour Cooperative Society",
      score: Math.round(score * 10) / 10,
      distance_km: Math.round(d * 10) / 10,
      avg_rating: w.avg_rating || 4.8,
      rating_count: w.rating_count || 12,
      experience_years: w.experience_years || 5.0,
      reasons
    });
  }

  if (out.length > 0) {
    return out.sort((a, b) => b.score - a.score).slice(0, limit);
  }

  // Fallback: State-wide Federation Hub Dispatch
  for (const w of workers) {
    if (w.verification_status !== "VERIFIED") continue;
    const wid = w.worker_id || w.id;
    if (!workerFree(wid, targetDate, startTime, durationMin)) continue;

    const wLat = 11.0168;
    const wLng = 76.9558;
    const d = haversineKm(lat, lng, wLat, wLng);
    const score = 35 + 20 + 10 + Math.min(10, ((w.avg_rating || 4.5) / 5) * 10) + (svc.requires_certification ? 5 : 0);

    const reasons = [
      "Required trade skill matches",
      "Schedule slot available",
      `State Cooperative Federation Hub Dispatch (${d.toFixed(0)} km transit zone)`,
      "Labour Cooperative verified tradesperson"
    ];
    if (svc.requires_certification) {
      reasons.push("Official trade certification verified");
    }
    if (w.avg_rating && w.avg_rating > 0) {
      reasons.push(`${w.avg_rating.toFixed(1)} ★ customer rating`);
    }

    out.push({
      worker_id: wid,
      name: w.name,
      phone: w.phone,
      cooperative_name: w.cooperative || "Tamil Nadu Labour Cooperative Federation",
      score: Math.round(score * 10) / 10,
      distance_km: Math.round(d * 10) / 10,
      avg_rating: w.avg_rating || 4.5,
      rating_count: w.rating_count || 10,
      experience_years: w.experience_years || 5.0,
      reasons
    });
  }

  return out.sort((a, b) => b.score - a.score).slice(0, limit);
}
